File: zac_js/src/flutter/services.ts
import { ZacConverter, ZacConverterType } from "../base";
import { ZacValue } from "../zac/zac_value";
import { Brightness, Color } from "./dart_ui";

export class SystemUiOverlayStyle extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static new(data: {
        systemNavigationBarColor?: Color,
        systemNavigationBarDividerColor?: Color,
        systemNavigationBarIconBrightness?: Brightness,
        systemNavigationBarContrastEnforced?: ZacValue<boolean>,
        statusBarColor?: Color,
        statusBarBrightness?: Brightness,
        statusBarIconBrightness?: Brightness,
        systemStatusBarContrastEnforced?: ZacValue<boolean>
    }) {
        return new SystemUiOverlayStyle({ converter: 'f:1:SystemUiOverlayStyle', ...data })
    }
}
export class TextInputType extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static datetime() {
        return new TextInputType({ converter: 'f:1:TextInputType.datetime' })
    }
    static emailAddress() {
        return new TextInputType({ converter: 'f:1:TextInputType.emailAddress' })
    }
    static multiline() {
        return new TextInputType({ converter: 'f:1:TextInputType.multiline' })
    }
    static name_() {
        return new TextInputType({ converter: 'f:1:TextInputType.name' })
    }
    static none() {
        return new TextInputType({ converter: 'f:1:TextInputType.none' })
    }
    static number() {
        return new TextInputType({ converter: 'f:1:TextInputType.number' })
    }
    static phone() {
        return new TextInputType({ converter: 'f:1:TextInputType.phone' })
    }
    static streetAddress() {
        return new TextInputType({ converter: 'f:1:TextInputType.streetAddress' })
    }
    static text() {
        return new TextInputType({ converter: 'f:1:TextInputType.text' })
    }
    static url() {
        return new TextInputType({ converter: 'f:1:TextInputType.url' })
    }
    static visiblePassword() {
        return new TextInputType({ converter: 'f:1:TextInputType.visiblePassword' })
    }
}
export class TextInputAction extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static continueAction() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.continueAction' })
    }
    static done() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.done' })
    }
    static emergencyCall() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.emergencyCall' })
    }
    static go() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.go' })
    }
    static join() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.join' })
    }
    static newline() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.newline' })
    }
    static next() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.next' })
    }
    static none() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.none' })
    }
    static previous() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.previous' })
    }
    static route() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.route' })
    }
    static search() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.search' })
    }
    static send() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.send' })
    }
    static unspecified() {
        return new TextInputAction({ converter: 'f:1:TextInputAction.unspecified' })
    }
}
export class TextCapitalization extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static characters() {
        return new TextCapitalization({ converter: 'f:1:TextCapitalization.characters' })
    }
    static none() {
        return new TextCapitalization({ converter: 'f:1:TextCapitalization.none' })
    }
    static sentences() {
        return new TextCapitalization({ converter: 'f:1:TextCapitalization.sentences' })
    }
    static words() {
        return new TextCapitalization({ converter: 'f:1:TextCapitalization.words' })
    }
}
export class SmartDashesType extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static disabled() {
        return new SmartDashesType({ converter: 'f:1:SmartDashesType.disabled' })
    }
    static enabled() {
        return new SmartDashesType({ converter: 'f:1:SmartDashesType.enabled' })
    }
}
export class SmartQuotesType extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static disabled() {
        return new SmartQuotesType({ converter: 'f:1:SmartQuotesType.disabled' })
    }
    static enabled() {
        return new SmartQuotesType({ converter: 'f:1:SmartQuotesType.enabled' })
    }
}
export class ScrollViewKeyboardDismissBehavior extends ZacConverter {
    private constructor(data: ZacConverterType) {
        super(data);
    }
    static manual() {
        return new ScrollViewKeyboardDismissBehavior({ converter: 'f:1:ScrollViewKeyboardDismissBehavior.manual' })
    }
    static onDrag() {
        return new ScrollViewKeyboardDismissBehavior({ converter: 'f:1:ScrollViewKeyboardDismissBehavior.onDrag' })
    }
}

File: zac/packages/zac/lib/src/flutter/services.dart
import 'package:flutter/widgets.dart';
import 'package:zac/src/zac/context.dart';
import 'package:zac/src/zac/zac_builder.dart';
import 'package:flutter/services.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:zac/src/base.dart';

part 'services.freezed.dart';
part 'services.g.dart';

@freezedZacBuilder
class FlutterSystemUiOverlayStyle
    with _$FlutterSystemUiOverlayStyle, ZacBuilder<SystemUiOverlayStyle> {
  const FlutterSystemUiOverlayStyle._();

  factory FlutterSystemUiOverlayStyle.fromJson(Map<String, dynamic> json) =>
      _$FlutterSystemUiOverlayStyleFromJson(json);

  @FreezedUnionValue('f:1:SystemUiOverlayStyle')
  factory FlutterSystemUiOverlayStyle({
    ZacBuilder<Color?>? systemNavigationBarColor,
    ZacBuilder<Color?>? systemNavigationBarDividerColor,
    ZacBuilder<Brightness?>? systemNavigationBarIconBrightness,
    ZacBuilder<bool?>? systemNavigationBarContrastEnforced,
    ZacBuilder<Color?>? statusBarColor,
    ZacBuilder<Brightness?>? statusBarBrightness,
    ZacBuilder<Brightness?>? statusBarIconBrightness,
    ZacBuilder<bool?>? systemStatusBarContrastEnforced,
  }) = _FlutterSystemUiOverlayStyle;

  SystemUiOverlayStyle _build(BuildContext context, ZacContext zacContext) {
    return SystemUiOverlayStyle(
      systemNavigationBarColor:
          systemNavigationBarColor?.build(context, zacContext),
      systemNavigationBarDividerColor:
          systemNavigationBarDividerColor?.build(context, zacContext),
      systemNavigationBarIconBrightness:
          systemNavigationBarIconBrightness?.build(context, zacContext),
      systemNavigationBarContrastEnforced:
          systemNavigationBarContrastEnforced?.build(context, zacContext),
      statusBarColor: statusBarColor?.build(context, zacContext),
      statusBarBrightness: statusBarBrightness?.build(context, zacContext),
      statusBarIconBrightness:
          statusBarIconBrightness?.build(context, zacContext),
      systemStatusBarContrastEnforced:
          systemStatusBarContrastEnforced?.build(context, zacContext),
    );
  }

  @override
  SystemUiOverlayStyle build(BuildContext context, ZacContext zacContext) {
    return _build(context, zacContext);
  }
}

@freezedZacBuilder
class FlutterTextInputType
    with _$FlutterTextInputType, ZacBuilder<TextInputType> {
  const FlutterTextInputType._();

  factory FlutterTextInputType.fromJson(Map<String, dynamic> json) =>
      _$FlutterTextInputTypeFromJson(json);

  @FreezedUnionValue('f:1:TextInputType.datetime')
  factory FlutterTextInputType.datetime() = _FlutterTextInputTypedatetime;
  @FreezedUnionValue('f:1:TextInputType.emailAddress')
  factory FlutterTextInputType.emailAddress() =
      _FlutterTextInputTypeemailAddress;
  @FreezedUnionValue('f:1:TextInputType.multiline')
  factory FlutterTextInputType.multiline() = _FlutterTextInputTypemultiline;
  @FreezedUnionValue('f:1:TextInputType.name')
  factory FlutterTextInputType.name() = _FlutterTextInputTypename;
  @FreezedUnionValue('f:1:TextInputType.none')
  factory FlutterTextInputType.none() = _FlutterTextInputTypenone;
  @FreezedUnionValue('f:1:TextInputType.number')
  factory FlutterTextInputType.number() = _FlutterTextInputTypenumber;
  @FreezedUnionValue('f:1:TextInputType.phone')
  factory FlutterTextInputType.phone() = _FlutterTextInputTypephone;
  @FreezedUnionValue('f:1:TextInputType.streetAddress')
  factory FlutterTextInputType.streetAddress() =
      _FlutterTextInputTypestreetAddress;
  @FreezedUnionValue('f:1:TextInputType.text')
  factory FlutterTextInputType.text() = _FlutterTextInputTypetext;
  @FreezedUnionValue('f:1:TextInputType.url')
  factory FlutterTextInputType.url() = _FlutterTextInputTypeurl;
  @FreezedUnionValue('f:1:TextInputType.visiblePassword')
  factory FlutterTextInputType.visiblePassword() =
      _FlutterTextInputTypevisiblePassword;

  TextInputType _build(BuildContext context, ZacContext zacContext) {
    return map(
      datetime: (_) => TextInputType.datetime,
      emailAddress: (_) => TextInputType.emailAddress,
      multiline: (_) => TextInputType.multiline,
      name: (_) => TextInputType.name,
      none: (_) => TextInputType.none,
      number: (_) => TextInputType.number,
      phone: (_) => TextInputType.phone,
      streetAddress: (_) => TextInputType.streetAddress,
      text: (_) => TextInputType.text,
      url: (_) => TextInputType.url,
      visiblePassword: (_) => TextInputType.visiblePassword,
    );
  }

  @override
  TextInputType build(BuildContext context, ZacContext zacContext) {
    return _build(context, zacContext);
  }
}

@freezedZacBuilder
class FlutterTextInputAction
    with _$FlutterTextInputAction, ZacBuilder<TextInputAction> {
  const FlutterTextInputAction._();

  factory FlutterTextInputAction.fromJson(Map<String, dynamic> json) =>
      _$FlutterTextInputActionFromJson(json);

  @FreezedUnionValue('f:1:TextInputAction.continueAction')
  factory FlutterTextInputAction.continueAction() =
      _FlutterTextInputActioncontinueAction;
  @FreezedUnionValue('f:1:TextInputAction.done')
  factory FlutterTextInputAction.done() = _FlutterTextInputActiondone;
  @FreezedUnionValue('f:1:TextInputAction.emergencyCall')
  factory FlutterTextInputAction.emergencyCall() =
      _FlutterTextInputActionemergencyCall;
  @FreezedUnionValue('f:1:TextInputAction.go')
  factory FlutterTextInputAction.go() = _FlutterTextInputActiongo;
  @FreezedUnionValue('f:1:TextInputAction.join')
  factory FlutterTextInputAction.join() = _FlutterTextInputActionjoin;
  @FreezedUnionValue('f:1:TextInputAction.newline')
  factory FlutterTextInputAction.newline() = _FlutterTextInputActionnewline;
  @FreezedUnionValue('f:1:TextInputAction.next')
  factory FlutterTextInputAction.next() = _FlutterTextInputActionnext;
  @FreezedUnionValue('f:1:TextInputAction.none')
  factory FlutterTextInputAction.none() = _FlutterTextInputActionnone;
  @FreezedUnionValue('f:1:TextInputAction.previous')
  factory FlutterTextInputAction.previous() = _FlutterTextInputActionprevious;
  @FreezedUnionValue('f:1:TextInputAction.route')
  factory FlutterTextInputAction.route() = _FlutterTextInputActionroute;
  @FreezedUnionValue('f:1:TextInputAction.search')
  factory FlutterTextInputAction.search() = _FlutterTextInputActionsearch;
  @FreezedUnionValue('f:1:TextInputAction.send')
  factory FlutterTextInputAction.send() = _FlutterTextInputActionsend;
  @FreezedUnionValue('f:1:TextInputAction.unspecified')
  factory FlutterTextInputAction.unspecified() =
      _FlutterTextInputActionunspecified;

  TextInputAction _build(BuildContext context, ZacContext zacContext) {
    return map(
      continueAction: (_) => TextInputAction.continueAction,
      done: (_) => TextInputAction.done,
      emergencyCall: (_) => TextInputAction.emergencyCall,
      go: (_) => TextInputAction.go,
      join: (_) => TextInputAction.join,
      newline: (_) => TextInputAction.newline,
      next: (_) => TextInputAction.next,
      none: (_) => TextInputAction.none,
      previous: (_) => TextInputAction.previous,
      route: (_) => TextInputAction.route,
      search: (_) => TextInputAction.search,
      send: (_) => TextInputAction.send,
      unspecified: (_) => TextInputAction.unspecified,
    );
  }

  @override
  TextInputAction build(BuildContext context, ZacContext zacContext) {
    return _build(context, zacContext);
  }
}

@freezedZacBuilder
class FlutterTextCapitalization
    with _$FlutterTextCapitalization, ZacBuilder<TextCapitalization> {
  const FlutterTextCapitalization._();

  factory FlutterTextCapitalization.fromJson(Map<String, dynamic> json) =>
      _$FlutterTextCapitalizationFromJson(json);

  @FreezedUnionValue('f:1:TextCapitalization.characters')
  factory FlutterTextCapitalization.characters() =
      _FlutterTextCapitalizationcharacters;
  @FreezedUnionValue('f:1:TextCapitalization.none')
  factory FlutterTextCapitalization.none() = _FlutterTextCapitalizationnone;
  @FreezedUnionValue('f:1:TextCapitalization.sentences')
  factory FlutterTextCapitalization.sentences() =
      _FlutterTextCapitalizationsentences;
  @FreezedUnionValue('f:1:TextCapitalization.words')
  factory FlutterTextCapitalization.words() = _FlutterTextCapitalizationwords;

  TextCapitalization _build(BuildContext context, ZacContext zacContext) {
    return map(
      characters: (_) => TextCapitalization.characters,
      none: (_) => TextCapitalization.none,
      sentences: (_) => TextCapitalization.sentences,
      words: (_) => TextCapitalization.words,
    );
  }

  @override
  TextCapitalization build(BuildContext context, ZacContext zacContext) {
    return _build(context, zacContext);
  }
}

@freezedZacBuilder
class FlutterSmartDashesType
    with _$FlutterSmartDashesType, ZacBuilder<SmartDashesType> {
  const FlutterSmartDashesType._();

  factory FlutterSmartDashesType.fromJson(Map<String, dynamic> json) =>
      _$FlutterSmartDashesTypeFromJson(json);

  @FreezedUnionValue('f:1:SmartDashesType.disabled')
  factory FlutterSmartDashesType.disabled() = _FlutterSmartDashesTypedisabled;

  @FreezedUnionValue('f:1:SmartDashesType.enabled')
  factory FlutterSmartDashesType.enabled() = _FlutterSmartDashesTypeenabled;

  SmartDashesType _build(BuildContext context, ZacContext zacContext) {
    return map(
      disabled: (_) => SmartDashesType.disabled,
      enabled: (_) => SmartDashesType.enabled,
    );
  }

  @override
  SmartDashesType build(BuildContext context, ZacContext zacContext) {
    return _build(context, zacContext);
  }
}

@freezedZacBuilder
class FlutterSmartQuotesType
    with _$FlutterSmartQuotesType, ZacBuilder<SmartQuotesType> {
  const FlutterSmartQuotesType._();

  factory FlutterSmartQuotesType.fromJson(Map<String, dynamic> json) =>
      _$FlutterSmartQuotesTypeFromJson(json);

  @FreezedUnionValue('f:1:SmartQuotesType.disabled')
  factory FlutterSmartQuotesType.disabled() = _FlutterSmartQuotesTypedisabled;

  @FreezedUnionValue('f:1:SmartQuotesType.enabled')
  factory FlutterSmartQuotesType.enabled() = _FlutterSmartQuotesTypeenabled;

  SmartQuotesType _build(BuildContext context, ZacContext zacContext) {
    return map(
      disabled: (_) => SmartQuotesType.disabled,
      enabled: (_) => SmartQuotesType.enabled,
    );
  }

  @override
  SmartQuotesType build(BuildContext context, ZacContext zacContext) {
    return _build(context, zacContext);
  }
}
